### PR TITLE
Implement support for iif()

### PR DESCRIPTION
**Pull request by thisismiller**
In sqlite, iif() looks like:

```
sqlite> create table iiftest(a int, b int, c int);
sqlite> explain select iif(a,b,c) from iiftest;
addr  opcode         p1    p2    p3    p4             p5  comment
----  -------------  ----  ----  ----  -------------  --  -------------
0     Init           0     11    0                    0   Start at 11
1     OpenRead       0     2     0     3              0   root=2 iDb=0; iiftest
2     Rewind         0     10    0                    0
3       Column         0     0     2                    0   r[2]= cursor 0 column 0
4       IfNot          2     7     1                    0
5       Column         0     1     1                    0   r[1]= cursor 0 column 1
6       Goto           0     8     0                    0
7       Column         0     2     1                    0   r[1]= cursor 0 column 2
8       ResultRow      1     1     0                    0   output=r[1]
9     Next           0     3     0                    1
10    Halt           0     0     0                    0
11    Transaction    0     0     1     0              1   usesStmtJournal=0
12    Goto           0     1     0                    0
```

And with this change, in limbo it looks like:

```
addr  opcode             p1    p2    p3    p4             p5  comment
----  -----------------  ----  ----  ----  -------------  --  -------
0     Init               0     14    0                    0   Start at 14
1     OpenReadAsync      0     2     0                    0   table=iiftest, root=2
2     OpenReadAwait      0     0     0                    0
3     RewindAsync        0     0     0                    0
4     RewindAwait        0     13    0                    0   Rewind table iiftest
5       Column           0     0     2                    0   r[2]=iiftest.a
6       IfNot            2     9     1                    0   if !r[2] goto 9
7       Column           0     1     1                    0   r[1]=iiftest.b
8       Goto             0     10    0                    0
9       Column           0     2     1                    0   r[1]=iiftest.c
10      ResultRow        1     1     0                    0   output=r[1]
11    NextAsync          0     0     0                    0
12    NextAwait          0     5     0                    0
13    Halt               0     0     0                    0
14    Transaction        0     0     0                    0
15    Goto               0     1     0                    0
```

---

Rust commentary overly welcome, as idk what I'm doing in this language.

### CODE REVIEW

File: core/translate/expr.rs [hunk 1]
```diff
@@ -967,6 +967,56 @@ pub fn translate_expr(
...
+                                &args[0],
+                                temp_reg,
+                                precomputed_exprs_to_registers,
+                            )?;
+                            let jump_target_when_false = program.allocate_label();
+                            program.emit_insn_with_label_dependency(
+                                Insn::IfNot {
+                                    reg: temp_reg,
+                                    target_pc: jump_target_when_false,
+                                    null_reg: 1,
```
Comment (thisismiller) [inline core/translate/expr.rs:991]: I don't quite understand what the null_reg is about, so this isn't right, but I'm not sure what is.

> If the value in P1 is NULL then take the jump if and only if P3 is non-zero.

Would make me think I'm supposed to allocate a register and load it with a non-zero constant for this, but the equivalent sqlite code just points at `1` which was never touched or initialized or anything.  So, I don't quite get it.

Comment (jussisaurio) [reply core/translate/expr.rs:991]: I don't precisely know when it's used with a 0 and non-0, but a high level comment here:https://github.com/penberg/limbo/pull/425#discussion_r1876253256

File: core/translate/expr.rs [hunk 2]
```diff
@@ -967,6 +967,56 @@ pub fn translate_expr(
...
+                                precomputed_exprs_to_registers,
+                            )?;
+                            let jump_target_result = program.allocate_label();
+                            program.emit_insn_with_label_dependency(
+                                Insn::Goto {
+                                    target_pc: jump_target_result,
+                                },
+                                jump_target_result,
+                            );
+                            program.preassign_label_to_next_insn(jump_target_when_false);
```
Comment (jussisaurio) [inline core/translate/expr.rs:1009]: you can just use `program.resolve_label(jump_target_when_false, program.offset())` here and below

File: testing/scalar-functions.test [hunk 3]
```diff
@@ -75,6 +75,14 @@ do_execsql_test ifnull-2 {
   select ifnull(null, 2);
 } {2}
 
+do_execsql_test iif-1 {
+  select iif(1, 1, 0);
```
[the remark this answers is not in the document]
Comment (thisismiller) [reply testing/scalar-functions.test:79]: I can flip to 'correct'/'incorrect' if you prefer, but isn't the point of these tests to test one thing?  I would have expected it'd be actively preferred to leave the expression and conditions as uncomplicated as possible, as then it keeps expression evaluation breakage to expression tests?

[the remark this answers is not in the document]
Comment (jussisaurio) [reply testing/scalar-functions.test:79]: Yeah I think you're right. You can leave it or change the 2nd and 3rd params, however you wanna do

**Review by jussisaurio — APPROVED** [review]
great stuff

noticed there's a merge conflict now, could you please resolve that and then i'll merge

**Comment by thisismiller** [comment]
merge conflict with last_insert_rowid fixed

**Pull request merged**